File: src/main/java/ru/darkkeks/vkmirror/autoreg/BotAutoReg.java
package ru.darkkeks.vkmirror.autoreg;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import ru.darkkeks.vkmirror.VkMirrorTelegram;
import ru.darkkeks.vkmirror.tdlib.TdApi;

import java.util.concurrent.BlockingQueue;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.ScheduledExecutorService;
import java.util.regex.Matcher;
import java.util.regex.Pattern;


public class BotAutoReg {

    private static final Logger logger = LoggerFactory.getLogger(BotAutoReg.class);

    private static final Pattern timeoutPattern = Pattern.compile("Please try again in (\\d+) seconds.");

    private static final String MY_BOTS = "/mybots";
    private static final String BOT_FATHER = "BotFather";

    public static final int NO_TIMEOUT = -1;


    private VkMirrorTelegram account;
    private ScheduledExecutorService executor;

    private BlockingQueue<BotFatherAction<?>> actions;
    private BlockingQueue<TdApi.Message> responses;

    private TdApi.Chat botFatherChat;

    public BotAutoReg(VkMirrorTelegram account, ScheduledExecutorService executor) {
        this.account = account;
        this.executor = executor;

        this.actions = new LinkedBlockingQueue<>();
        this.responses = new LinkedBlockingQueue<>();

        account.onMessage(updateNewMessage -> {
            TdApi.Message message = updateNewMessage.message;
            if (message.chatId == botFatherChat.id && message.senderUserId != account.getMyId()) {
                responses.offer(message);
            }
        });

        account.searchPublicUsername(BOT_FATHER).thenAccept(chat -> {
            botFatherChat = chat;

            executor.submit(() -> {
                // It is here to reset state if it is messed up for some reason
                executeCommand(MY_BOTS);

                runActionFromQueue();
            });
        });
    }

    public String executeCommand(String command) {
        account.sendMessage(botFatherChat.id, command);
        try {
            // BotFather shouldn't reply with anything that is not text
            return ((TdApi.MessageText) responses.take().content).text.text;
        } catch (InterruptedException e) {
            logger.error("Interrupted while waiting for " + command + " response");
            throw new IllegalStateException(e);
        }
    }

    private void runActionFromQueue() {
        try {
            BotFatherAction<?> action = actions.take();
            executor.submit(() -> {
                action.execute();
                runActionFromQueue();
            });
        } catch (InterruptedException e) {
            logger.error("Interrupted while waiting for action", e);
        }
    }

    /**
     * @param response Response from BotFather supposedly containing timeout message
     * @return timeout in seconds, or {@value NO_TIMEOUT} is non is present
     */
    public int getTimeoutSeconds(String response) {
        Matcher matcher = timeoutPattern.matcher(response);
        if (matcher.find()) {
            return Integer.parseInt(matcher.group(1));
        }
        return NO_TIMEOUT;
    }

    /**
     * Creates new bot with specified name and username
     * @param username New bot username (without @), has to not be used
     * @param name New bot name
     * @return CompletableFuture with new bot token
     */
    public CompletableFuture<String> register(String username, String name) {
        RegisterAction action = new RegisterAction(this, username, name);
        actions.offer(action);
        return action.getCompletableFuture();
    }

    /**
     * Sets bot description
     * @param username Bot username (with @)
     * @param description New description
     */
    public CompletableFuture<Void> setDescription(String username, String description) {
        SetDescriptionAction action = new SetDescriptionAction(this, username, description);
        actions.offer(action);
        return action.getCompletableFuture();
    }

    // TODO BotAutoReg::setAvatar
    public CompletableFuture<Void> setAvatar(String username, Object avatar) {
        throw new UnsupportedOperationException();
    }

}

File: src/main/java/ru/darkkeks/vkmirror/bot/BotAutoReg.kt
package ru.darkkeks.vkmirror.bot

import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.launch
import org.kodein.di.Kodein
import org.kodein.di.generic.instance
import ru.darkkeks.vkmirror.tdlib.TelegramClient
import ru.darkkeks.vkmirror.tdlib.internal.TdApi


class BotAutoReg(kodein: Kodein) {

    val client: TelegramClient by kodein.instance()

    private val actions = Channel<BotFatherAction<*>>()
    private val responses = Channel<TdApi.Message>()

    private lateinit var botFatherChat: TdApi.Chat

    suspend fun start() {
        client.subscribe<TdApi.UpdateNewMessage> {
            val message = it.message
            if (message.chatId == botFatherChat.id && message.senderUserId != client.myId) {
                println("Received message $message")
                GlobalScope.launch {
                    // FIXME
                    responses.send(message)
                }
            }
        }

        botFatherChat = client.searchPublicUsername(BOT_FATHER)

        // Reset state if it is messed up for some reason
        executeCommand(START)

        actionLoop()
    }

    private suspend fun actionLoop() {
        while (true) {
            val action = actions.receive()
            action.execute()
        }
    }

    suspend fun executeCommand(command: String): String {
        sendMessage(command)
        val response = responses.receive()
        val content = response.content
        return (content as TdApi.MessageText).text.text
    }

    suspend fun sendMessage(message: String) = client.sendMessage(botFatherChat.id, message)

    /**
     * Creates new bot with specified name and username
     * @param username Bot username (without @), has to not be used
     * @param name Bot name
     * @return registered bot token
     */
    suspend fun register(username: String, name: String): String {
        val action = RegisterAction(this, username, name)
        actions.send(action)
        return action.callback.await()
    }

    /**
     * Sets bot about text
     * @param username Bot username (with @)
     * @param about New about text
     */
    suspend fun setAbout(username: String, about: String) {
        val action = SetAboutAction(this, username, about)
        actions.send(action)
        action.callback.await()
    }

    /**
     * Sets bot name
     * @param username Bot username (with @)
     * @param name New bot name
     */
    suspend fun setName(username: String, name: String) {
        val action = SetNameAction(this, username, name)
        actions.send(action)
        action.callback.await()
    }

    // TODO BotAutoReg::setAvatar
    suspend fun setAvatar(username: String, avatar: Any) {
        throw UnsupportedOperationException()
    }
}
